scripts/docs_link_checker.py

```diff
@@ -0,0 +1,306 @@
+
+        Args:
+            match: A positive match of a markdown link (ex: [...](...))
+            file: The file where the match was found
+
+        Returns:
+            A LinkReport if the link is broken, otherwise None
+        """
+        link = match.group(2)
+
+        # skip links that are anchor only (start with #)
+        if self._is_anchor_link(link):
+            return None
+
+        if self._external_link_pattern.match(link):
+            result = self._check_external_link(link, file)
+        elif self._is_image_link(match.group(0)):
+            match = self._relative_image_pattern.match(link)
+            if match:
+                result = self._check_relative_image(link, file, match.group("path"))
+            else:
+                match = self._absolute_image_pattern.match(link)
+                if match:
+                    result = self._check_absolute_image(link, file, match.group("path"))
+                else:
+                    result = LinkReport(link, file, "Invalid image link format")
+        else:
+            match = self._relative_link_pattern.match(link)
+            if match:
+                result = self._check_relative_link(link, file, match.group("path"))
+            else:
+                match = self._absolute_link_pattern.match(link)
+                if match:
+                    result = self._check_absolute_link(link, file, match.group("path"))
+                else:
+                    match = self._docroot_link_pattern.match(link)
+                    if match:
+                        result = self._check_docroot_link(
+                            link, file, match.group("path")
+                        )
+                    else:
+                        result = LinkReport(link, file, "Invalid link format")
+
+        return result
+
+    def check_file(self, file: str) -> List[LinkReport]:
+        """Looks for all the links in a file and checks them.
+
+        Returns:
+            A list of broken links, or an empty list if no links are broken
+        """
+        with open(file) as f:
+            contents = f.read()
+
+        matches = self._markdown_link_pattern.finditer(contents)
+
+        result: List[LinkReport] = []
+
+        for match in matches:
+            report = self.check_link(match, file)
+            if report is not None:
+                result.append(report)
+
+        return result
+
+
+@click.command()
+@click.option("--path", "-p", default=".", help="Path to markdown files to check")
+@click.option(
+    "--docs-root", "-r", default=None, help="Root to all docs for link checking"
+)
+@click.option(
+    "--site-prefix",
+    "-s",
+    default="",
+    help="Top-most folder in the docs URL for resolving absolute paths",
+)
+@click.option("--skip-external", is_flag=True)
+def scan_docs(path: str, docs_root: str, site_prefix: str, skip_external: bool):
+    # verify that our path is correct
```

Comment:
```suggestion
def scan_docs(path: str, docs_root: str, site_prefix: str, skip_external: bool) -> None:
    # verify that our path is correct
```
I believe untyped returns are implicit `Any` so just being explicit

Reply:
Also `docs_root` is `Optional` right?

Reply:
yes, typehint updated